File: examples/canonical-faker/javascript/src/validate.ts
import type { Dataset, ValidationReport } from "./domain.js"
import {
  apiKeyIds,
  BASE_TIME,
  buildPools,
  fromIso,
  invitationIds,
  membershipIds,
  organizationById,
  userById
} from "./pools.js"

export function validateDataset(dataset: Dataset): ValidationReport {
  const violations: string[] = []
  const counts: Record<string, number> = {
    organizations: dataset.organizations.length,
    users: dataset.users.length,
    memberships: dataset.memberships.length,
    projects: dataset.projects.length,
    audit_events: dataset.auditEvents.length,
    api_keys: dataset.apiKeys.length,
    invitations: dataset.invitations.length
  }
  const minimumRows: Record<string, number> = {
    organizations: 1,
    users: 1,
    memberships: 1,
    projects: 1,
    audit_events: 1,
    api_keys: 0,
    invitations: 0
  }
  for (const [entity, count] of Object.entries(counts)) {
    if (count < minimumRows[entity]!) {
      violations.push(`row count below minimum for ${entity}: ${count} < ${minimumRows[entity]}`)
    }
  }

  const orgs = organizationById(dataset.organizations)
  const users = userById(dataset.users)
  const pools = buildPools(dataset)
  const projectIds = new Set(dataset.projects.map((row) => row.id))
  const membershipIdSet = membershipIds(dataset.memberships)
  const apiKeyIdSet = apiKeyIds(dataset.apiKeys)
  const invitationIdSet = invitationIds(dataset.invitations)

  const seenOrgIds = new Set<string>()
  const seenEmails = new Set<string>()
  const seenPrefixes = new Set<string>()
  const memberEmails = new Map<string, Set<string>>()

  for (const org of dataset.organizations) {
    if (seenOrgIds.has(org.id)) {
      violations.push(`duplicate organizations.id: ${org.id}`)
    }
    seenOrgIds.add(org.id)
  }
  for (const user of dataset.users) {
    const email = user.email.toLowerCase()
    if (seenEmails.has(email)) {
      violations.push(`duplicate user.email: ${email}`)
    }
    seenEmails.add(email)
  }
  for (const membership of dataset.memberships) {
    const org = orgs.get(membership.org_id)
    const user = users.get(membership.user_id)
    if (!org) {
      violations.push(`membership missing org: ${membership.id}`)
      continue
    }
    if (!user) {
      violations.push(`membership missing user: ${membership.id}`)
      continue
    }
    const emails = memberEmails.get(membership.org_id) ?? new Set<string>()
    emails.add(user.email.toLowerCase())
    memberEmails.set(membership.org_id, emails)
    if (fromIso(membership.joined_at) < fromIso(org.created_at)) {
      violations.push(`Rule A violated by membership ${membership.id}`)
    }
    if (membership.invited_by !== null && !users.has(membership.invited_by)) {
      violations.push(`membership invited_by missing user: ${membership.id}`)
    }
  }
  for (const project of dataset.projects) {
    const org = orgs.get(project.org_id)
    if (!org) {
      violations.push(`project missing org: ${project.id}`)
      continue
    }
    if (fromIso(project.created_at) < fromIso(org.created_at)) {
      violations.push(`Rule B violated by project ${project.id}`)
    }
    if (!(pools.orgMemberMap.get(project.org_id) ?? []).includes(project.created_by)) {
      violations.push(`Rule C violated by project ${project.id}`)
    }
  }
  for (const event of dataset.auditEvents) {
    const org = orgs.get(event.org_id)
    const project = pools.projectById.get(event.project_id)
    if (!org) {
      violations.push(`audit event missing org: ${event.id}`)
      continue
    }
    if (!project) {
      violations.push(`audit event missing project: ${event.id}`)
      continue
    }
    if (!(pools.orgMemberMap.get(event.org_id) ?? []).includes(event.user_id)) {
      violations.push(`Rule D violated by audit event ${event.id}`)
    }
    if (project.org_id !== event.org_id) {
      violations.push(`Rule E violated by audit event ${event.id}`)
    }
    const eventTs = fromIso(event.ts)
    if (eventTs < fromIso(project.created_at)) {
      violations.push(`Rule F violated by audit event ${event.id}`)
    }
    const membership = pools.membershipByOrgUser.get(`${event.org_id}:${event.user_id}`)
    if (membership && eventTs < fromIso(membership.joined_at)) {
      violations.push(`audit event before membership joined_at: ${event.id}`)
    }
    if (event.resource_type === "project" && !projectIds.has(event.resource_id)) {
      violations.push(`audit event resource project missing: ${event.id}`)
    }
    if (event.resource_type === "user" && !users.has(event.resource_id)) {
      violations.push(`audit event resource user missing: ${event.id}`)
    }
    if (event.resource_type === "membership" && !membershipIdSet.has(event.resource_id)) {
      violations.push(`audit event resource membership missing: ${event.id}`)
    }
    if (event.resource_type === "api_key" && !apiKeyIdSet.has(event.resource_id)) {
      violations.push(`audit event resource api_key missing: ${event.id}`)
    }
    if (event.resource_type === "invitation" && !invitationIdSet.has(event.resource_id)) {
      violations.push(`audit event resource invitation missing: ${event.id}`)
    }
  }
  for (const apiKey of dataset.apiKeys) {
    if (!(pools.orgMemberMap.get(apiKey.org_id) ?? []).includes(apiKey.created_by)) {
      violations.push(`Rule G violated by api_key ${apiKey.id}`)
    }
    if (seenPrefixes.has(apiKey.key_prefix)) {
      violations.push(`duplicate api_key.key_prefix: ${apiKey.key_prefix}`)
    }
    seenPrefixes.add(apiKey.key_prefix)
    if (apiKey.last_used_at !== null && fromIso(apiKey.last_used_at) < fromIso(apiKey.created_at)) {
      violations.push(`api_key last_used_at before created_at: ${apiKey.id}`)
    }
  }
  for (const invitation of dataset.invitations) {
    if (!(pools.orgMemberMap.get(invitation.org_id) ?? []).includes(invitation.invited_by)) {
      violations.push(`Rule H violated by invitation ${invitation.id}`)
    }
    if ((memberEmails.get(invitation.org_id) ?? new Set<string>()).has(invitation.invited_email.toLowerCase())) {
      violations.push(`Rule I violated by invitation ${invitation.id}`)
    }
    if (fromIso(invitation.expires_at) <= BASE_TIME) {
      violations.push(`invitation expiry must be in the future: ${invitation.id}`)
    }
    if (invitation.accepted_at !== null && fromIso(invitation.accepted_at) > BASE_TIME) {
      violations.push(`invitation accepted_at must be in the past: ${invitation.id}`)
    }
  }

  return {
    ok: violations.length === 0,
    violations,
    counts,
    seed: dataset.seed,
    profileName: dataset.profileName
  }
}

File: examples/canonical-faker/rust/src/validate.rs
use std::collections::{HashMap, HashSet};

use chrono::Duration;
use serde::Serialize;

use crate::domain::{Dataset, Membership};
use crate::pools::{base_time, build_graph};

#[derive(Debug, Clone, Serialize)]
pub struct ValidationReport {
    pub ok: bool,
    pub violations: Vec<String>,
    pub counts: HashMap<String, usize>,
    pub seed: u64,
    pub profile_name: String,
}

pub fn validate(dataset: &Dataset) -> ValidationReport {
    let mut violations = Vec::new();
    let counts = HashMap::from([
        ("organizations".to_string(), dataset.organizations.len()),
        ("users".to_string(), dataset.users.len()),
        ("memberships".to_string(), dataset.memberships.len()),
        ("projects".to_string(), dataset.projects.len()),
        ("audit_events".to_string(), dataset.audit_events.len()),
        ("api_keys".to_string(), dataset.api_keys.len()),
        ("invitations".to_string(), dataset.invitations.len()),
    ]);
    let graph = build_graph(dataset);

    let orgs: HashMap<_, _> = dataset
        .organizations
        .iter()
        .map(|org| (org.id.clone(), org))
        .collect();
    let users: HashMap<_, _> = dataset.users.iter().map(|user| (user.id.clone(), user)).collect();
    let projects: HashMap<_, _> = dataset
        .projects
        .iter()
        .map(|project| (project.id.clone(), project))
        .collect();

    let mut org_ids = HashSet::new();
    for org in &dataset.organizations {
        if !org_ids.insert(org.id.clone()) {
            violations.push(format!("duplicate organizations.id: {}", org.id));
        }
    }

    let mut user_emails = HashSet::new();
    for user in &dataset.users {
        if !user_emails.insert(user.email.to_lowercase()) {
            violations.push(format!("duplicate users.email: {}", user.email));
        }
    }

    let mut membership_ids = HashSet::new();
    for membership in &dataset.memberships {
        membership_ids.insert(membership.id.clone());
        let Some(org) = orgs.get(&membership.org_id) else {
            violations.push(format!("membership missing org: {}", membership.id));
            continue;
        };
        if !users.contains_key(&membership.user_id) {
            violations.push(format!("membership missing user: {}", membership.id));
            continue;
        }
        if membership.joined_at < org.created_at {
            violations.push(format!("Rule A violated by membership {}", membership.id));
        }
        if let Some(invited_by) = &membership.invited_by {
            if !users.contains_key(invited_by) {
                violations.push(format!("membership invited_by missing user: {}", membership.id));
            }
        }
    }

    let mut api_key_ids = HashSet::new();
    let mut key_prefixes = HashSet::new();
    for api_key in &dataset.api_keys {
        api_key_ids.insert(api_key.id.clone());
        if !graph
            .org_member_map
            .get(&api_key.org_id)
            .is_some_and(|members| members.contains(&api_key.created_by))
        {
            violations.push(format!("Rule G violated by api_key {}", api_key.id));
        }
        if !key_prefixes.insert(api_key.key_prefix.clone()) {
            violations.push(format!("duplicate api_key.key_prefix: {}", api_key.key_prefix));
        }
        if let Some(last_used) = api_key.last_used_at {
            if last_used < api_key.created_at {
                violations.push(format!("api_key last_used_at before created_at: {}", api_key.id));
            }
        }
    }

    let mut invitation_ids = HashSet::new();
    for invitation in &dataset.invitations {
        invitation_ids.insert(invitation.id.clone());
        if !graph
            .org_member_map
            .get(&invitation.org_id)
            .is_some_and(|members| members.contains(&invitation.invited_by))
        {
            violations.push(format!("Rule H violated by invitation {}", invitation.id));
        }
        if graph
            .org_member_map
            .get(&invitation.org_id)
            .into_iter()
            .flatten()
            .filter_map(|member_id| graph.user_email_by_id.get(member_id))
            .any(|email| email == &invitation.invited_email.to_lowercase())
        {
            violations.push(format!("Rule I violated by invitation {}", invitation.id));
        }
        if invitation.expires_at <= base_time() {
            violations.push(format!("invitation expiry must be in the future: {}", invitation.id));
        }
        if invitation.expires_at > base_time() + Duration::days(30) {
            violations.push(format!("invitation expiry must be within 30 days: {}", invitation.id));
        }
        if invitation
            .accepted_at
            .is_some_and(|accepted_at| accepted_at > base_time())
        {
            violations.push(format!("invitation accepted_at must be in the past: {}", invitation.id));
        }
    }

    for project in &dataset.projects {
        let Some(org) = orgs.get(&project.org_id) else {
            violations.push(format!("project missing org: {}", project.id));
            continue;
        };
        if project.created_at < org.created_at {
            violations.push(format!("Rule B violated by project {}", project.id));
        }
        if !graph
            .org_member_map
            .get(&project.org_id)
            .is_some_and(|members| members.contains(&project.created_by))
        {
            violations.push(format!("Rule C violated by project {}", project.id));
        }
    }

    for event in &dataset.audit_events {
        let Some(project) = projects.get(&event.project_id) else {
            violations.push(format!("audit_event missing project: {}", event.id));
            continue;
        };
        if !orgs.contains_key(&event.org_id) {
            violations.push(format!("audit_event missing org: {}", event.id));
            continue;
        }
        if !graph
            .org_member_map
            .get(&event.org_id)
            .is_some_and(|members| members.contains(&event.user_id))
        {
            violations.push(format!("Rule D violated by audit_event {}", event.id));
        }
        if project.org_id != event.org_id {
            violations.push(format!("Rule E violated by audit_event {}", event.id));
        }
        if event.ts < project.created_at {
            violations.push(format!("Rule F violated by audit_event {}", event.id));
        }
        if let Some(membership) = graph
            .membership_by_org_user
            .get(&format!("{}:{}", event.org_id, event.user_id))
        {
            if event.ts < membership.joined_at {
                violations.push(format!("audit event before membership joined_at: {}", event.id));
            }
        }
        match event.resource_type.as_str() {
            "project" if !projects.contains_key(&event.resource_id) => {
                violations.push(format!("audit event resource project missing: {}", event.id));
            }
            "user" if !users.contains_key(&event.resource_id) => {
                violations.push(format!("audit event resource user missing: {}", event.id));
            }
            "membership" if !membership_ids.contains(&event.resource_id) => {
                violations.push(format!("audit event resource membership missing: {}", event.id));
            }
            "api_key" if !api_key_ids.contains(&event.resource_id) => {
                violations.push(format!("audit event resource api_key missing: {}", event.id));
            }
            "invitation" if !invitation_ids.contains(&event.resource_id) => {
                violations.push(format!("audit event resource invitation missing: {}", event.id));
            }
            _ => {}
        }
    }

    ValidationReport {
        ok: violations.is_empty(),
        violations,
        counts,
        seed: dataset.seed,
        profile_name: dataset.profile_name.clone(),
    }
}

#[allow(dead_code)]
fn membership_is_zero(membership: &Membership) -> bool {
    membership.id.is_empty()
}
